File: javascript/ecdsa/security/audit/ecdsa-signature-forgery.js
var ellipticcurve = require("starkbank-ecdsa");
var Ecdsa = ellipticcurve.Ecdsa;
var PrivateKey = ellipticcurve.PrivateKey;

// Generate new Keys
let privateKey = new PrivateKey();
let publicKey = privateKey.publicKey();

let message = "My test message";

// Generate Signature
let signature = Ecdsa.sign(message, privateKey);

// Verify if signature is valid
// ruleid: ecdsa-signature-forgery
console.log(Ecdsa.verify(message, signature, publicKey));

var ellipticecurve2 = require("starkbank-ecdsa");
var Ecdsa2 = ellipticecurve2.Ecdsa;
var Signature = ellipticecurve2.Signature;
var PublicKey = ellipticecurve2.PublicKey;
var File = ellipticecurve2.utils.File;

let publicKeyPem = File.read("publicKey.pem");
let signatureDer = File.read("signatureDer.txt", "binary");
let message2 = File.read("message.txt");

let publicKey2 = PublicKey.fromPem(publicKeyPem);
let signature2 = Signature.fromDer(signatureDer);

// ruleid: ecdsa-signature-forgery
console.log(Ecdsa2.verify(message2, signature2, publicKey2));

File: python/cryptography/security/ecdsa-signature-forgery.py
from ellipticcurve.ecdsa import Ecdsa
from ellipticcurve.privateKey import PrivateKey

# Generate new Keys
privateKey = PrivateKey()
publicKey = privateKey.publicKey()

message = "My test message"

# Generate Signature
signature = Ecdsa.sign(message, privateKey)

# To verify if the signature is valid
# ruleid: ecdsa-signature-forgery
Ecdsa.verify(message, signature, publicKey)

publicKeyPem = File.read("publicKey.pem")
signatureDer = File.read("signatureDer.txt", "rb")
message2 = File.read("message.txt")

publicKey2 = PublicKey.fromPem(publicKeyPem)
signature2 = Signature.fromDer(signatureDer)

# ruleid: ecdsa-signature-forgery
print(Ecdsa.verify(message2, signature2, publicKey2))
